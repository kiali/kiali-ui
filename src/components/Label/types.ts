export interface Labels {
  name: string;
  value: string;
}
