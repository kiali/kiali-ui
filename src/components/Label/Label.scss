.label-collection {
  overflow: hidden;
}

.label-pair {
  margin: 3px 2px 2px;
  max-width: 100%;
  float: left;
  display: block;

  .label {
    min-width: 50px;

    &:first-child {
      border-radius: 2px 0 0 2px;
    }

    &:last-child {
      border-radius: 0 2px 2px 0;
    }
  }

  .label-key {
    background-color: #bee1f4;
    color: #00659c;
    padding-top: 4px;
    border-radius: 2px 0 0 2px;
    padding-bottom: 4px;
    margin-left: 0;
  }

  .label-value {
    background-color: #7dc3e8;
    color: #fff;
    padding-top: 4px;
    border-radius: 0 2px 2px 0;
    padding-bottom: 4px;
    margin-left: 0;
  }
}
