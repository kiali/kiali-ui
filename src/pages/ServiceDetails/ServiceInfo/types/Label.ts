export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
