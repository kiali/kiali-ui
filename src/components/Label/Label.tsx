import * as React from 'react';
import { Label as PfLabel } from 'patternfly-react';
import './Label.css';
import { canRender } from '../../utils/SafeRender';

interface Props {
  name: string;
  value: string;
}

const Label = (props: Props) => {
  const { name, value } = props;

  if (canRender(name) && canRender(value)) {
    return (
      <span className="label-pair">
        <PfLabel bsStyle="primary" className="label-key">
          {name}
        </PfLabel>
        <PfLabel bsStyle="primary" className="label-value">
          {value || ''}
        </PfLabel>
      </span>
    );
  } else {
    return <span>This label has an unexpected format</span>;
  }
};

export default Label;
