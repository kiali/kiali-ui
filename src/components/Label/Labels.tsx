import * as React from 'react';
import Label from './Label';
import { style } from 'typestyle';

const SHOW_MORE_TRESHOLD = 2;

interface Props {
  labels?: { [key: string]: string };
}

interface State {
  expanded: boolean;
}

const linkStyle = style({
  float: 'left',
  margin: '7px 2px 2px 3px',
  fontSize: '0.8rem'
});

class Labels extends React.Component<Props, State> {
  constructor(props: Props, state: State) {
    super(props, state);
    this.state = {
      expanded: false
    };
  }

  labelKeys() {
    return Object.keys(this.props.labels || {});
  }

  hasLabels() {
    return this.labelKeys().length > 0;
  }

  hasManyLabels() {
    return this.labelKeys().length > SHOW_MORE_TRESHOLD;
  }

  showItem(i: number) {
    return this.state.expanded || !this.hasManyLabels() || i < SHOW_MORE_TRESHOLD;
  }

  expandLabels = () => {
    this.setState({ expanded: true });
  };

  renderMoreLabelsLink() {
    if (this.hasManyLabels() && !this.state.expanded) {
      return (
        <a key="label_more" className={linkStyle} onClick={this.expandLabels}>
          {' '}
          More labels...
        </a>
      );
    }

    return null;
  }

  renderLabels() {
    return this.labelKeys().map((key, i) => {
      const hideClass = this.showItem(i) ? '' : 'hide';
      return (
        <div key={'label_' + i} className={hideClass}>
          <Label key={'label_' + i} name={key} value={this.props.labels ? this.props.labels[key] : ''} />
        </div>
      );
    });
  }

  renderEmptyLabels() {
    return <span> No labels </span>;
  }

  render() {
    if (this.hasLabels()) {
      return [this.renderLabels(), this.renderMoreLabelsLink()];
    } else {
      return this.renderEmptyLabels();
    }
  }
}

export default Labels;
