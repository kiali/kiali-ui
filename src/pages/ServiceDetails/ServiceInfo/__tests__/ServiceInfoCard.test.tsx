import * as React from 'react';
import { shallow } from 'enzyme';
import Label from '../../../../components/Label/Label';

const CardContent = (
  <div key="pod">
    <div>
      <strong>Pod:</strong>
    </div>
    <ul style={{ listStyleType: 'none' }}>
      <li key="pod_labels_badge_">
        <Label name="my_key" value="my_value" />
      </li>
    </ul>
  </div>
);

describe('#PfInfoCard render correctly with data', () => {
  it('should render service card', () => {
    const wrapper = shallow(
      <div className="card-pf">
        <div className="card-pf-body">{CardContent}</div>
      </div>
    );
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
