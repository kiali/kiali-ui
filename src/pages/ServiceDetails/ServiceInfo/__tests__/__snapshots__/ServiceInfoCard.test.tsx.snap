// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#PfInfoCard render correctly with data should render service card 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div
    className="card-pf"
  >
    <div
      className="card-pf-body"
    >
      <div>
        <div>
          <strong>
            Pod:
          </strong>
        </div>
        <ul
          style={
            Object {
              "listStyleType": "none",
            }
          }
        >
          <li>
            <Label
              name="my_key"
              value="my_value"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="card-pf-body"
      >
        <div>
          <div>
            <strong>
              Pod:
            </strong>
          </div>
          <ul
            style={
              Object {
                "listStyleType": "none",
              }
            }
          >
            <li>
              <Label
                name="my_key"
                value="my_value"
              />
            </li>
          </ul>
        </div>
      </div>,
      "className": "card-pf",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div>
            <strong>
              Pod:
            </strong>
          </div>
          <ul
            style={
              Object {
                "listStyleType": "none",
              }
            }
          >
            <li>
              <Label
                name="my_key"
                value="my_value"
              />
            </li>
          </ul>
        </div>,
        "className": "card-pf-body",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": "pod",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              <strong>
                Pod:
              </strong>
            </div>,
            <ul
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            >
              <li>
                <Label
                  name="my_key"
                  value="my_value"
                />
              </li>
            </ul>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <strong>
                Pod:
              </strong>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Pod:",
              },
              "ref": null,
              "rendered": "Pod:",
              "type": "strong",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <li>
                <Label
                  name="my_key"
                  value="my_value"
                />
              </li>,
              "style": Object {
                "listStyleType": "none",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "pod_labels_badge_",
              "nodeType": "host",
              "props": Object {
                "children": <Label
                  name="my_key"
                  value="my_value"
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "name": "my_key",
                  "value": "my_value",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "li",
            },
            "type": "ul",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="card-pf-body"
        >
          <div>
            <div>
              <strong>
                Pod:
              </strong>
            </div>
            <ul
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            >
              <li>
                <Label
                  name="my_key"
                  value="my_value"
                />
              </li>
            </ul>
          </div>
        </div>,
        "className": "card-pf",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <div>
              <strong>
                Pod:
              </strong>
            </div>
            <ul
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            >
              <li>
                <Label
                  name="my_key"
                  value="my_value"
                />
              </li>
            </ul>
          </div>,
          "className": "card-pf-body",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "pod",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                <strong>
                  Pod:
                </strong>
              </div>,
              <ul
                style={
                  Object {
                    "listStyleType": "none",
                  }
                }
              >
                <li>
                  <Label
                    name="my_key"
                    value="my_value"
                  />
                </li>
              </ul>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <strong>
                  Pod:
                </strong>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Pod:",
                },
                "ref": null,
                "rendered": "Pod:",
                "type": "strong",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <li>
                  <Label
                    name="my_key"
                    value="my_value"
                  />
                </li>,
                "style": Object {
                  "listStyleType": "none",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "pod_labels_badge_",
                "nodeType": "host",
                "props": Object {
                  "children": <Label
                    name="my_key"
                    value="my_value"
                  />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "name": "my_key",
                    "value": "my_value",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "li",
              },
              "type": "ul",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
